scripts/qa_crowbarsetup.sh

```diff
@@ -4478,6 +4618,107 @@ function oncontroller
     run_on "$novacontroller" "oncontroller_$func $@"
 }
 
+function oncompute
+{
+    local func=$1 ; shift
+    run_on "$novacompute_kvm" "oncompute_$func $@"
+}
+
+function oncompute_add_intel_iommu_param
+{
+    cat /etc/default/grub | grep iommu >/dev/null 2>&1
+    if [ $? != 0 ]; then
+        # enabled intel_iommu in kernel param
+        sed -in "s/\(^GRUB_CMDLINE_LINUX_DEFAULT=.*\)\"$/\1 intel_iommu=on\"/g" /etc/default/grub
+        update-bootloader
+    fi
+}
+
+function oncompute_load_vfio_module_onboot
+{
+    if [ ! -e "/etc/modules-load.d/pci-passthru.conf" ]; then
+        # vfio-pci module is not loaded by default, however, we need it to bind to
+        # our candidate device to verify pci passthru
+        cat > /etc/modules-load.d/pci-passthru.conf <<EOF
+vfio_pci
+EOF
+        modprobe vfio-pci
+    fi
+}
+
+function oncompute_bind_vfio_module_onboot
+{
+    if [ ! -e "/etc/udev/rules.d/99-pci-passthru.rules" ]; then
+            # We need to bind vfio-pci to our candidate device to verify pci passthru
+            cat > /etc/udev/rules.d/99-pci-passthru.rules <<EOF
+KERNELS=="0000:00:1d.0", SUBSYSTEMS=="pci", ATTRS{device}=="0x1001", ATTRS{vendor}=="0x1af4", \
+RUN+="/bin/bash -c 'echo  0000:00:1d.0 > /sys/bus/pci/devices/0000:00:1d.0/driver/unbind'", \
+RUN+="/bin/bash -c 'echo 1af4 1001 > /sys/bus/pci/drivers/vfio-pci/new_id'"
+EOF
+    fi
+}
+
+function oncompute_create_pci_passthru_conf
+{
+    if [ ! -e "/etc/nova/nova.conf.d/200-nova-pci-passthru.conf" ]; then
+        # drop new configuration file under /etc/nova/nova.conf.d/ on compute
+        # [pci]
+        # passthrough_whitelist = { "address": "0000:41:00.0" }
+        # alias = { "vendor_id":"1af4", "product_id":"1001", "device_type":"type-PCI", "name":"a1" }
+        cat > /etc/nova/nova.conf.d/200-nova-pci-passthru.conf <<EOF
+[pci]
+passthrough_whitelist = { "address": "0000:00:1d.0" }
+alias = { "vendor_id":"1af4", "product_id":"1001", "device_type":"type-PCI", "name":"a1" }
+EOF
+        # Make the changes effective
+        systemctl restart openstack-nova-compute.service
+    fi
+}
+
+function oncontroller_create_pci_passthru_conf
+{
+    if [ ! -e "/etc/nova/nova.conf.d/200-nova-pci-passthru.conf" ]; then
+        # drop new configuration file under /etc/nova/nova.conf.d/ on controller
+        # [pci]
+        # alias = { "vendor_id":"1af4", "product_id":"1001", "device_type":"type-PCI", "name":"a1" }
+        cat > /etc/nova/nova.conf.d/200-nova-pci-passthru.conf <<EOF
+[pci]
+alias = { "vendor_id":"1af4", "product_id":"1001", "device_type":"type-PCI", "name":"a1" }
+EOF
+        # Make the changes effective
+        systemctl restart openstack-nova-api.service
+    fi
+}
+
+function append_pci_passthru_filter
```

Comment:
This function goes to a lot of trouble to just set one parameter in the nova proposal. Instead, why not add that parameter from the very beginning, when the proposal gets first created [here](https://github.com/SUSE-Cloud/automation/blob/master/scripts/qa_crowbarsetup.sh#L2612-L2667) ?

Reply:
It's part of the acceptance criteria for this [Jira](https://jira.suse.com/browse/SOC-9304) which is not introducing any change in Nova proposal for testing pci passthrough but sneak in the configuration when it is needed.